### ISSUE TITLE
Panic in `apSubst`

### ISSUE THREAD

**Issue by abakst** [GaloisInc/cryptol#916]
I'm not sure if this is "my" error or not, as I came across this panic while using Cryptol as a library (hence the pretty-printed term below etc).

Given this term:
```
\{n, a} (fin i`961, fin n, n >= i`961) (x : [n]a) ->
(Cryptol::#) i`961 (n - i`961) a <> (Cryptol::drop (n -
                                                    i`961) i`961 a <> x) (Cryptol::take (n -
                                                                                         i`961) i`961 a <> x)
```
if I apply the substitution
```
i`961 = 2
```
then cryptol panics:
```
  Message:   Predicate split or disappeared after
             we applied a substitution.
             Predicate:
             fin i`961
             Became:
             []
             subst:
             Substitution:
```

The full message is: 
```
%< ---------------------------------------------------
  Revision:  f0483684081711e1f698e1d7c933395f74d57eb4
  Branch:    master
  Location:  apSubst@EProofAbs
  Message:   Predicate split or disappeared after
             we applied a substitution.
             Predicate:
             fin i`961
             Became:
             []
             subst:
             Substitution:
  i`961 = 2
CallStack (from HasCallStack):
  panic, called at src/Cryptol/Utils/Panic.hs:21:9 in cryptol-2.9.1.99-inplace:Cryptol.Utils.Panic
  panic, called at src/Cryptol/TypeCheck/Subst.hs:358:32 in cryptol-2.9.1.99-inplace:Cryptol.TypeCheck.Subst
%< ---------------------------------------------------
```
My hunch is that this error is cropping up _now_ due to some additional strictness: I reverted to an older commit to compare behaviors and could not reproduce until I replaced the `apSubst sub e` with `apSubst sub (traceShow (pp e) e)`

**Comment by yav** [comment]
Yes, this was probably always a problem, but maybe it was hidden in a thunk and the recent changes to avoid memory leaks probably just brought it to light.

Generally, `apSubst` is intended for substituting types for *free* type variables, and the assumptions is that these are only around while we are checking a specific function. 

If you are working with the IR, you could use a type application to instantiate a polymorphic value, otherwise maybe jump on the `cryptol` channel and we can see what other options there are.

**Comment by yav** [comment]
I'll close this for now, as I don't think we need to fix anything.  Please feel free to reopen if you think otherwise.

**Comment by brianhuffman** [comment]
I don't think we necessarily need to change the code, but it's probably a good policy that uses of Cryptol-as-a-library should never panic unless the calling code has violated a documented precondition or datatype invariant. So we should document the precondition here before closing. (I'm actually not even sure what the precondition is, exactly.)

**Comment by yav** [comment]
I think of `apSubst` as being internal to the type checker, and wasn't thinking if it as being part of the API.   It is mostly intended to substitute types for free type variables, which shouldn't exist outside of the type checker.   

The original intention was that the API to the typechecker is just the module `Cryptol.TypeCheck` + the relevant AST.  Maybe we can make this more explicitly when we rearrange the modules.

**Comment by brianhuffman** [comment]
It would be nice to document the precondition anyway, even if it's just for internal benefit :)
